### PR TITLE
Lower warp_execute_on_lane_0 to scf.if

### PR DESCRIPTION
**Pull request by matthias-springer**
This transform assumes that the first argument to the enclosing function
has a memref buffer that can be used as scratch pad memory.

Only rank 1 vectors are supported.

### CODE REVIEW

**Comment by matthias-springer** [comment]
Note: Pull request has 2 commits for easier reviewing.

File: lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp [hunk 1]
```diff
@@ -263,3 +264,144 @@ void mlir::vector_ext::populatePropagateVectorDistributionPatterns(
...
+    // Compute next offset.
+    rewriter.setInsertionPoint(ifOp);
+    Value vectorSize = rewriter.create<arith::ConstantIndexOp>(
+        yieldLoc, vectorType.getShape()[0]);
+    storeOffset = rewriter.create<arith::AddIOp>(yieldLoc, storeOffset,
+                                                 vectorSize);
+  }
+
+  // Delete terminator and add empty scf.yield.
+  rewriter.eraseOp(yieldOp);
```
Comment (ThomasRaoux) [inline lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp:375]: nit: is this needed since we are replacing the whole warpOp anyway?

**Review by ThomasRaoux — APPROVED** [review]
Thanks Matthias! This looks great, added few nits, we will need to make temp memory shared memory allocs to be able to run this code.

File: lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp [hunk 2]
```diff
@@ -263,3 +264,144 @@ void mlir::vector_ext::populatePropagateVectorDistributionPatterns(
...
+    bbArgReplacements.push_back(loadOp);
+
+    // Compute next offset.
+    rewriter.setInsertionPoint(ifOp);
+    loadOffset = rewriter.create<arith::AddIOp>(loc, loadOffset,
+        rewriter.create<arith::ConstantIndexOp>(loc, bbArgType.getShape()[0]));
+  }
+
+  // Move body of warpOp to ifOp.
+  rewriter.mergeBlocks(warpOpBody, ifOp.thenBlock(), bbArgReplacements);
```
[the remark this answers is not in the document]
Comment (matthias-springer) [reply lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp:336]: The load is in fact created directly in the `ifOp` block

File: lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp (hunk 1 above)
Comment (matthias-springer) [reply lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp:375]: The yieldOp was moved over to the `scf.if` by `mergeBlocks` and must be deleted.

File: lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp [hunk 3]
```diff
@@ -263,3 +264,144 @@ void mlir::vector_ext::populatePropagateVectorDistributionPatterns(
...
+                                PatternRewriter &rewriter) const override {
+    auto funcOp = warpOp->getParentOfType<FuncOp>();
+    if (!funcOp)
+      return failure();
+
+    // We assume that the first FuncOp argument is a memref buffer that can be
+    // used as scratch pad memory.
+    if (funcOp.getArgumentTypes().empty() ||
+        !funcOp.getArgumentTypes()[0].isa<MemRefType>())
+      return failure();
```
[the remark this answers is not in the document]
Comment (matthias-springer) [reply lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp:397]: Not sure what address space constant to use. Just creating a normal `memref.alloc` for now.

**Pull request merged**